### DISCUSSION TITLE
Federation setup for Matrix

### DISCUSSION THREAD

**Discussion by gouthamravee** [NginxProxyManager/nginx-proxy-manager#2404] [Q&A]
I'm trying to get federation working for my matrix server. I'm using the ansible install method. 
Matrix and nginx-proxy are on different vms with their own unique address on the LAN. 

I have the ansible nginx-proxy on but using different ports, and NPM is managing proxying from `matrix.my.domain` and `my.domain`. 
NPM is also managing certs for both domains. 
https://github.com/spantaleev/matrix-docker-ansible-deploy/blob/master/docs/configuring-playbook-own-webserver.md#method-2-fronting-the-integrated-nginx-reverse-proxy-webserver-with-another-reverse-proxy

I can access matrix from matrix.my.domain. 
I have port `8448` forwarded to NPM and I have the dockerfile setup to pass port `8448` to `8048` on the matrix server. 

The problem is I need `my.domain:8448` to show the SSL cert for my.domain

In NPM I have hosts with SSL created for 
```
matrix.my.domain -> matrix.my.lan:8008 
my.domain -> matrix.my.lan:8084
```

I also have a stream host as 
`8448 -> matrix.my.lan:8084`

But I'm getting a connection refused when doing the matrix federation test.

**Comment by marrobHD** [comment]
Hi,

i'd recommend to stick with port 443. Ther is absolutely no need to forward other ports than 80&443 for NPM.
No need to create a stream host.
Just create two proxy hosts:
- my.domain -> web-server-with-well-known:internal-port
- matrix.my.domain -> your-matrix-server:internal-matrix-synapse-port
You need to create a well-known for my.domain for matrix, because it doesn't try on port 443. Please take a look into spec: [/.well-known/matrix/server](https://matrix.org/docs/spec/server_server/r0.1.0.html#get-well-known-matrix-server)

After that you should be able to test your instance with https://federationtester.matrix.org/#my.domain

**Reply by untenops** [reply]
I know this is old, but I'm trying to do this exact thing and have no idea how to setup, well-known.  I can't seem to find anything on well-known with Nginx Proxy Manager.  
Could someone explain this to me like I'm 5
Thanks

**Reply by michael-oe** [reply]
Hello untenops,

here is my setup, which works for me:

- matrix-server (docker) installed and running under
  
  - Domain: matrix.DOMAIN.TLD (DOMAIN.TLD has to be replaced with your domain name!)
    
  - local ip-adress: 192.xxx.xxx.xxx (has to replace with your IP for matrix server)
    
  - port 8008
    
- For federation I placed 2 text-files (client, server) on a web server at DOMAIN.TLD, reachable from the internet under following adresses (here DOMAIN.TLD not matrix.DOMAIN.TLD)
  
  - https://domain.tld/.well-known/matrix/client
    with content:
    {
     "m.homeserver": {"base_url": "[https://matrix.DOMAIN.TLD"}](https://matrix.DOMAIN.TLD"%7D)
     "m.identity_server": {"base_url": "[https://vector.im"}](https://vector.im"%7D)
    }
    
  - https://domain.tld/.well-known/matrix/server
    with content:
    {
     "m.server": "matrix.DOMAIN.TLD:443"
    }
    
- NPM-Settings (proxy host):
  
  - Details:
    
    - Domain Names: matrix.DOMAIN.TLD
      
    - Scheme: http
      
    - Forward Hostname/IP: 192.xxx.xxx.xxx
      
    - Forward Port: 8008
      
    - activated: Block Common Exploits, Websockets Support
      
  - Advanced:
    
    - listen 8448 ssl http2 default_server;
      listen [::]:8448 ssl http2 default_server;
      location /_synapse/admin {
       return 404;
      }
      
    - NOTICE: the part "location /_synapse/admin { return 404; }" disables the admin interface of synapse for security reasons; if you want to use the admin interface delete this lines or comment it out!
      
  - SSL:
    
    - request SSL certificate for matrix.DOMAIN.TLD
      
- Port Forwarding on router:
  
  - Port 8448 TCP and UDP has to be forwarded to your Matrix-Server-IP (not NPM IP if different)

Hope, this will help you.
Michael

**Reply by untenops** [reply]
Thank you.  I think this is what I needed.  I will try it out later today.

**Reply by untenops** [reply]
Thank you again.  After much banging of the head I finally got it to work.  Not sure, but seemed to start working after adding 

serve_server_wellknown: true

to the homeserver.yaml file.

**Reply by michael-oe** [reply]
Hello untenops,

If the Federation Test on ([https://federationtester.matrix.org/](https://federationtester.matrix.org/))  is successful, everything is fine.

My homeserver.yaml does not have a line "serve_server_wellknown: true". However, I have also installed the matrix server in Docker and the differences can certainly be explained there.

I wish you a lot of fun with your matrix server.
Best regards from Germany.
Michael
